cpp/src/dictionary/replace.cu

```diff
@@ -0,0 +1,192 @@
+  auto input_pair_iterator = make_nullable_index_iterator<true>(input);
+  return cudf::detail::copy_if_else(true,
+                                    input_pair_iterator,
+                                    input_pair_iterator + input.size(),
+                                    replacement_iter,
+                                    predicate,
+                                    mr,
+                                    stream);
+}
+
+}  // namespace
+
+/**
+ * @copydoc cudf::dictionary::detail::replace_nulls(cudf::column_view const&,cudf::column_view
+ * const&,rmm::mr::device_memory_resource*,cudaStream_t)
+ */
+std::unique_ptr<column> replace_nulls(dictionary_column_view const& input,
+                                      dictionary_column_view const& replacement,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+{
+  if (input.size() == 0) { return cudf::empty_like(input.parent()); }
+  if (!input.has_nulls()) { return std::make_unique<cudf::column>(input.parent()); }
+  CUDF_EXPECTS(input.keys().type() == replacement.keys().type(), "keys must match");
+  CUDF_EXPECTS(replacement.size() == input.size(), "column sizes must match");
+
+  // first combine the keys so both input dictionaries have the same set
+  auto matched = match_dictionaries({input, replacement}, mr, stream);
+
+  // now build the new indices by doing replace-null using the updated input indices
+  auto const input_indices =
+    dictionary_column_view(matched.front()->view()).get_indices_annotated();
+  auto const repl_indices = dictionary_column_view(matched.back()->view()).get_indices_annotated();
+  auto new_indices =
+    repl_indices.has_nulls()
+      ? replace_indices(input_indices, make_nullable_index_iterator<true>(repl_indices), mr, stream)
+      : replace_indices(
+          input_indices, make_nullable_index_iterator<false>(repl_indices), mr, stream);
+
+  // auto keys_column = ;
+  return make_dictionary_column(
+    std::move(matched.front()->release().children.back()), std::move(new_indices), mr, stream);
+}
+
+/**
+ * @copydoc cudf::dictionary::detail::replace_nulls(cudf::column_view const&,cudf::scalar
+ * const&,rmm::mr::device_memory_resource*,cudaStream_t)
+ */
+std::unique_ptr<column> replace_nulls(dictionary_column_view const& input,
+                                      scalar const& replacement,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+{
+  if (input.size() == 0) { return cudf::empty_like(input.parent()); }
+  if (!input.has_nulls() || !replacement.is_valid()) {
+    return std::make_unique<cudf::column>(input.parent());
+  }
+  CUDF_EXPECTS(input.keys().type() == replacement.type(), "keys must match scalar type");
+
+  // first add the replacment to the keys so only the indices need to be processed
+  auto const default_mr = rmm::mr::get_current_device_resource();
```

Comment:
Wait, can't you use match_dictionaries by using `make_column_from_scalar` first?

Reply:
`make_column_from_scalar` makes a column and does not make a dictionary. I think it would be wasteful to convert the scalar to a dictionary first.

Reply:
My whole point in suggesting what became `match_dictionaries` was to remove this duplication...

Reply:
I guess it's not duplication anymore because you used match_dictionaries in one of the two places. :)

Reply:
The `matched_dictionaries` will be very handy in other functions -- some of which have already been code.